diskqueue.go

```diff
@@ -395,6 +422,188 @@ func (d *diskQueue) readOne() ([]byte, error) {
+		}
+
+		return nil
+	}
+
+	err := d.walkDiskQueueDir(getAllBadFileInfo)
+
+	return badFileInfos, err
+}
+
+// get the accurate total non-"bad" file size
+func (d *diskQueue) updateTotalDiskSpaceUsed() error {
+	d.totalDiskSpaceUsed = maxMetaDataFileSize
+
+	updateTotalDiskSpaceUsed := func(fileInfo os.FileInfo) error {
+		// only accept files created by this DiskQueue object
+		if fileNameRegexp.MatchString(fileInfo.Name()) || badFileNameRegexp.MatchString(fileInfo.Name()) {
+			d.totalDiskSpaceUsed += fileInfo.Size()
+		}
+
+		return nil
+	}
+
+	return d.walkDiskQueueDir(updateTotalDiskSpaceUsed)
+}
+
+func (d *diskQueue) freeDiskSpace(expectedBytesIncrease int64) error {
+	var err error
+	var badFileInfos []os.FileInfo
+
+	badFileInfos, err = d.getAllBadFileInfo()
+	if err != nil {
+		d.logf(ERROR, "DISKQUEUE(%s) failed to retrieve all .bad file info - %s", d.name, err)
+	}
+
+	// keep freeing up disk space until we have enough space to write this message
+	for d.totalDiskSpaceUsed+expectedBytesIncrease > d.maxBytesDiskSpace {
+		if len(badFileInfos) > 0 {
+			// check if a .bad file exists. If it does, delete that first
+			oldestBadFileInfo := badFileInfos[0]
+			badFileFilePath := path.Join(d.dataPath, oldestBadFileInfo.Name())
+
+			err = os.Remove(badFileFilePath)
+			if err != nil {
+				d.logf(ERROR, "DISKQUEUE(%s) failed to remove .bad file(%s) - %s", d.name, oldestBadFileInfo.Name(), err)
+				d.updateTotalDiskSpaceUsed()
+				return err
+			} else {
+				// recaclulate total bad files disk size to get most accurate info
+				d.totalDiskSpaceUsed -= oldestBadFileInfo.Size()
+				d.logf(INFO, "DISKQUEUE(%s) removed .bad file(%s) of size(%d bytes) to free up disk space", d.name, oldestBadFileInfo.Name(), oldestBadFileInfo.Size())
+			}
+
+			badFileInfos = badFileInfos[1:]
+		} else {
+			// delete the read file (make space)
+			readFileToDeleteNum := d.readFileNum
+			err = d.removeReadFile()
+			if err != nil {
+				d.logf(ERROR, "DISKQUEUE(%s) failed to remove file(%s) - %s", d.name, d.fileName(readFileToDeleteNum), err)
+				d.handleReadError()
+				return err
+			} else {
+				d.logf(INFO, "DISKQUEUE(%s) removed file(%s) to free up disk space", d.name, d.fileName(readFileToDeleteNum))
+			}
+			d.updateTotalDiskSpaceUsed()
+		}
+	}
+
+	return nil
+}
+
+// check if there is enough available disk space to write new data to file
+func (d *diskQueue) checkDiskSpace(totalBytes int64, reachedFileSizeLimit bool) error {
+	expectedBytesIncrease := totalBytes
+	if reachedFileSizeLimit {
+		expectedBytesIncrease += numFileMsgBytes
+	}
+
+	// If the data to be written is bigger than the disk size limit, do not write
+	if expectedBytesIncrease > d.maxBytesDiskSpace {
+		errorMsg := fmt.Sprintf(
+			"message size(%d) surpasses disk size limit(%d)",
+			expectedBytesIncrease, d.maxBytesDiskSpace)
+		d.logf(ERROR, "DISKQUEUE(%s) - %s", d.name, errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	if d.maxBytesDiskSpace <= maxMetaDataFileSize {
```

Comment:
Could we move it to pre-check when creating diskqueue?

Reply:
This sounds like a great idea, but the `New` and `NewWithDiskSpace` functions only return an Interface and not an error. I made `NewWithDiskSpace` to have the same function signature as `New`. Do you recommend that I change the `NewWithDiskSpace` function so it returns: `Interface, error`? https://github.com/kev1n80/go-diskqueue/blob/DiskSizeLimit/diskqueue.go#L122-L138

Reply:
I prefer keeping same interface as before, 'Interface'. Could we return nil if it doesn't pass precheck?

Reply:
Wouldn't it be more descriptive to return an error when a user calls `diskQueue.Put()` instead of returning nil in the constructor? This should help the user debug the problem easier in a complex environment, instead of assuming that the nil object represents an error and why. Also, currently the original DiskQueue allows the user to construct a DiskQueue object with faulty inputs i.e. minMsgSize > maxMsgSize is valid.

Reply:
Yes, `Interface, error` would be better.

For precheck, I think it's needed.  Users could know it wouldn't work ASAP. The program would raise an error when it's starting.

Reply:
Hahaha agreed (not all users know d.maxBytesDiskSpace should be…) that’s why I thought having an error statement to explain how they can configure the constructor better in order to properly use it would be more effective than them guessing what needs to be fixed. However, I will add the precheck along with log statements.